File: 207-course_schedule.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        def build_graph():
            g = {}
            for i in range(numCourses):
                g.setdefault(i, set())
            for a, b in prerequisites:
                g[b].add(a)
            return g

        def dfs(graph, node, gray, black):
            nonlocal cycle
            if node in black:
                return

            if node in gray:
                cycle = True
                return

            gray.add(node)
            for nbr in graph[node]:
                dfs(graph, nbr, gray, black)

            gray.remove(node)
            black.add(node)


        graph = build_graph()
        gray, black = set(), set()
        cycle = False
        for c in range(numCourses):
            dfs(graph, c, gray, black)
            if cycle:
                return False

        return len(black) == numCourses

File: 207-course_schedule.cs
public class Solution {
    public bool CanFinish(int numCourses, int[][] prerequisites)
    {
        // build graph
        Dictionary<int, HashSet<int>> graph = new();
        for (int i = 0; i < numCourses; i++)
        {
            graph[i] = new();
        }
        foreach(int[] prereq in prerequisites)
        {
            var (a, b) = Tuple.Create(prereq[0], prereq[1]);
            graph[b].Add(a);
        }

        HashSet<int> gray = new();
        HashSet<int> black = new();
        bool cycle = false;

        for (int i = 0; i < numCourses; i++)
        {
            dfs(i);
            if (cycle)
            {
                return false;
            }
        }

        return true;

        void dfs(int node)
        {
            if (black.Contains(node))
            {
                return;  // already visited
            }

            if (gray.Contains(node))
            {
                cycle = true;
                return;
            }

            gray.Add(node);

            foreach(int nbr in graph[node]) {
                dfs(nbr);
            }

            gray.Remove(node);
            black.Add(node);
        }

        return false;
    }
}
